### PR TITLE
feat(api): replace `_provider_id` property with label-based isolation and regex injection for custom queries

### PR DESCRIPTION
**Pull request by josema-xyz**
### Description

- Replace the `_provider_id` property with `_Provider_{uuid}` label-based isolation across the Attack Paths module.
- Predefined queries no longer need explicit property filters: the Internet node is reached via `CAN_ACCESS` path connectivity, and `ProwlerFinding` drops its redundant `provider_uid` filter.
- Custom queries get automatic provider scoping through a new regex-based label injector (`cypher_rewriter.py`).
- Post-query filtering, subgraph operations, and schema lookups all switch from property checks to label checks.
- Two unused indexes are removed.

### Steps to review

Run tests and a local scan on an AWS provider using the UI. Then run Attack Paths pre-defined queries and some custom ones.

### Checklist

- [x] Review if the code is being covered by tests.
- [x] Review if code is being documented following this specification https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings
- [x] Review if backport is needed.
- [x] Review if is needed to change the [README.md](https://github.com/prowler-cloud/prowler/blob/master/README.md)
- [x] Ensure new entries are added to [CHANGELOG.md](https://github.com/prowler-cloud/prowler/blob/master/prowler/CHANGELOG.md), if applicable.

#### API
- [x] All issue/task requirements work as expected on the API
- [ ] Endpoint response output (if applicable)
- [ ] EXPLAIN ANALYZE output for new/modified queries or indexes (if applicable)
- [ ] Performance test results (if applicable)
- [ ] Any other relevant evidence of the implementation (if applicable)
- [x] Verify if API specs need to be regenerated.
- [x] Check if version updates are required (e.g., specs, Poetry, etc.).
- [x] Ensure new entries are added to [CHANGELOG.md](https://github.com/prowler-cloud/prowler/blob/master/api/CHANGELOG.md), if applicable.

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: api/src/backend/tasks/jobs/attack_paths/queries.py [hunk 1]
```diff
@@ -62,7 +61,6 @@ def render_cypher_template(template: str, replacements: dict[str, str]) -> str:
             finding.check_title = finding_data.check_title,
             finding.muted = finding_data.muted,
             finding.muted_reason = finding_data.muted_reason,
-            finding.provider_uid = $provider_uid,
             finding.firstseen = timestamp(),
             finding.lastupdated = $last_updated,
             finding._module_name = 'cartography:prowler',
```
[the remark this answers is not in the document]
Comment (josema-xyz) [reply api/src/backend/tasks/jobs/attack_paths/queries.py:66]: Not an issue. ProwlerFinding nodes are created in the temp database. During sync, `sync_nodes()` applies the `_Provider_{uuid}` label to all nodes (including findings) when merging into the tenant database. The label is there by the time the API serializes.

File: api/src/backend/tasks/jobs/attack_paths/queries.py [hunk 2]
```diff
@@ -84,7 +81,7 @@ def render_cypher_template(template: str, replacements: dict[str, str]) -> str:
 """
 
 CLEANUP_FINDINGS_TEMPLATE = f"""
-    MATCH (finding:{PROWLER_FINDING_LABEL} {{provider_uid: $provider_uid}})
+    MATCH (finding:{PROWLER_FINDING_LABEL})
         WHERE finding.lastupdated < $last_updated
 
     WITH finding LIMIT $batch_size
```
[the remark this answers is not in the document]
Comment (josema-xyz) [reply api/src/backend/tasks/jobs/attack_paths/queries.py:87]: Not a cross-provider risk. `CLEANUP_FINDINGS_TEMPLATE` runs in the temp database, which only contains data from one provider. The temp DB is dropped after sync.

That said, the `provider_uid` parameter was still being passed from `cleanup_findings()` even though the template no longer uses it. Removed that dead parameter.

File: api/src/backend/api/attack_paths/database.py [hunk 3]
```diff
@@ -163,24 +163,21 @@ def drop_subgraph(database: str, provider_id: str) -> int:
...
 
     try:
         with get_session(database) as session:
             deleted_count = 1
             while deleted_count > 0:
                 result = session.run(
                     f"""
-                    MATCH (n:{PROVIDER_RESOURCE_LABEL} {{{PROVIDER_ID_PROPERTY}: $provider_id}})
+                    MATCH (n:{PROVIDER_RESOURCE_LABEL}:`{provider_label}`)
                     WITH n LIMIT $batch_size
```
[the remark this answers is not in the document]
Comment (josema-xyz) [reply api/src/backend/api/attack_paths/database.py:176]: `provider_id` always comes from Django UUID model fields, so it's already validated by the time it reaches `_normalize_uuid`. No user-controlled input reaches this path. Adding UUID parsing here breaks existing tests that use simple string IDs, and the security gain is zero given the Django ORM guarantee.

File: api/src/backend/api/attack_paths/queries/schema.py [hunk 4]
```diff
@@ -1,13 +1,18 @@
-from tasks.jobs.attack_paths.config import PROVIDER_ID_PROPERTY, PROVIDER_RESOURCE_LABEL
+from tasks.jobs.attack_paths.config import PROVIDER_RESOURCE_LABEL, get_provider_label
+
+
+def get_cartography_schema_query(provider_id: str) -> str:
+    """Build the Cartography schema metadata query scoped to a provider label."""
+    provider_label = get_provider_label(provider_id)
+    return f"""
+        MATCH (n:{PROVIDER_RESOURCE_LABEL}:`{provider_label}`)
+        WHERE n._module_name STARTS WITH 'cartography:'
```
[the remark this answers is not in the document]
Comment (josema-xyz) [reply api/src/backend/api/attack_paths/queries/schema.py:9]: Same as above, covered by the Django UUID field guarantee.

File: api/src/backend/tasks/jobs/attack_paths/sync.py [hunk 5]
```diff
@@ -119,13 +119,7 @@ def sync_nodes(
...
                     NODE_SYNC_TEMPLATE, {"__NODE_LABELS__": node_labels}
                 )
-                target_session.run(
-                    query,
-                    {
-                        "rows": batch,
-                        "provider_id": provider_id,
-                    },
-                )
+                target_session.run(query, {"rows": batch})
```
[the remark this answers is not in the document]
Comment (josema-xyz) [reply api/src/backend/tasks/jobs/attack_paths/sync.py:117]: Fixed. Updated the docstring to reflect label-based isolation. Also updated the `scan.py` module docstring which had the same stale reference.

File: api/src/backend/tasks/jobs/attack_paths/sync.py [hunk 6]
```diff
@@ -174,13 +168,7 @@ def sync_relationships(
...
                     RELATIONSHIP_SYNC_TEMPLATE, {"__REL_TYPE__": rel_type}
                 )
-                target_session.run(
-                    query,
-                    {
-                        "rows": batch,
-                        "provider_id": provider_id,
-                    },
-                )
+                target_session.run(query, {"rows": batch})
```
[the remark this answers is not in the document]
Comment (josema-xyz) [reply api/src/backend/tasks/jobs/attack_paths/sync.py:165]: Fixed alongside the `sync_nodes` docstring.

File: api/src/backend/api/attack_paths/views_helpers.py [hunk 7]
```diff
@@ -174,6 +145,7 @@ def execute_custom_query(
     provider_id: str,
 ) -> dict[str, Any]:
     validate_custom_query(cypher)
+    cypher = inject_provider_label(cypher, provider_id)
```
Comment (AdriiiPRodri) [inline api/src/backend/api/attack_paths/views_helpers.py:155]: Can we add a comment explaining the 3 layers custom queries, something like:


```
# Defense-in-depth for custom queries:
# 1. neo4j.READ_ACCESS — prevents mutations at the driver level 
# 2. inject_provider_label() — regex-based label injection scopes node patterns
# 3. _serialize_graph() — post-query filter drops nodes without the provider label
#
# Layer 2 is best-effort (regex can't fully parse Cypher);
# layer 3 is the safety net that guarantees provider isolation.
```

File: api/src/backend/api/tests/test_cypher_sanitizer.py [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+
+class TestEdgeCases:
+    def test_empty_query(self):
+        assert _inject("") == ""
+
+    def test_no_node_patterns(self):
+        cypher = "RETURN 1 + 2"
+        assert _inject(cypher) == cypher
+
+    def test_anonymous_empty_parens_not_injected(self):
```
Comment (AdriiiPRodri) [inline api/src/backend/api/tests/test_cypher_sanitizer.py:307]: Can we add a test for the full-anonymous case? And document why it's safe

This test covers the case where one () is in a pattern alongside a labeled node. But the worst-case scenario, a query with only anonymous nodes isn't tested:

`MATCH ()--()--() RETURN *`

**Review by AdriiiPRodri — CHANGES_REQUESTED** [review]
Great job! 🚀

File: api/src/backend/api/attack_paths/views_helpers.py (hunk 7 above)
Comment (josema-xyz) [reply api/src/backend/api/attack_paths/views_helpers.py:155]: Added the defense-in-depth comment as suggested.

File: api/src/backend/api/tests/test_cypher_sanitizer.py (hunk 8 above)
Comment (josema-xyz) [reply api/src/backend/api/tests/test_cypher_sanitizer.py:307]: Added `test_fully_anonymous_query_bypasses_injection` covering the worst-case `MATCH ()--()--() RETURN *`. The test verifies injection leaves the query untouched and the docstring explains why that's safe: `_serialize_graph()`, the layer 3, drops any node missing the provider label before it reaches the user.

**Review by AdriiiPRodri — APPROVED** [review]
